### PR TITLE
Filter out CA PrivateKeyEntry when creating a KeyManager

### PR DESCRIPTION
**Pull request by BigPandaToo**
In 8.0, with security on by default, we store the HTTP
layer CA PrivateKeyEntry in the  http.ssl keystore (along
with the node certificate) so that it is available in our 
Enrollment API transport actions. 
When loading a keystore, the current behavior is that the
X509ExtendedKeyManager will iterate through the PrivateKeyEntry
objects and will return the first key/certificate that satisfies
the requirements of the client and the server configuration, 
and lacks any additional logic/filters.
We need the KeyManager to deterministically pick the node
certificate/key in all cases as this is the intended entry to be
used for TLS on the HTTP layer. 
This change introduces filtering when creating the in-memory
keystore the KeyManager is loaded with, so that it will not 
include PrivateKeyEntry objects when: 

- there are more than 1 PrivateKeyEntry objects in the keystore 
- The leaf certificate associated with the PrivateKeyEntry is a
CA certificate

Related: #75097

### CODE REVIEW

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by jkakavas** [comment]
@BigPandaToo I looked into the failing tests and these are actually bugs in the enroll node and enroll kibana APIs that we missed  originally because of lack of integration tests. I fixed them here instead of opening a new PR, let me know if you have any concerns with that

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [inline client/rest-high-level/qa/ssl-enabled/build.gradle:26]: Do we need these?

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [inline client/rest-high-level/qa/ssl-enabled/build.gradle:32]: Can we pick these up from the same place ?

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [inline client/rest-high-level/qa/ssl-enabled/build.gradle:39]: Why do we need verification mode to be none for transport ?

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [inline client/rest-high-level/qa/ssl-enabled/build.gradle:35]: We need to sync these settings (which options are set) with the ones we will be generating in https://github.com/elastic/elasticsearch/pull/74868. Not an action item to change something now necessarily, but a mental note to come back and do when #74868 is finalized/merged.

File: client/rest-high-level/qa/ssl-enabled/src/javaRestTest/java/org/elasticsearch/client/EnrollmentIT.java
Comment (jkakavas) [inline client/rest-high-level/qa/ssl-enabled/src/javaRestTest/java/org/elasticsearch/client/EnrollmentIT.java:70]: because of https://github.com/elastic/elasticsearch/pull/73807/files#diff-b33f30f231f98a863f3bd573ef551fe06c7e9904f957b3648b6d98d7e1a08ee9R24

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:102]: `ChangePasswordRequestBuilder#password` closes the SecureString that is passed to it, so we would always return a zeroed out char array otherwise

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:83]: Do we need `this instanceof StoreKeyConfig` ? We're in `StoreKeyConfig` right ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:93]: let's do `checkKeyStore(ks)` before we iterate the aliases for CA certificates

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:85]: we need to only be doing this for PrivateKeyEntry objects, not generally. This code will be removing TrustedCertificateEntry objects ( which _are_ CA certificates ) from the keystore and we do not want that. You can filter the aliases further with `ks.isKeyEntry(alias)` ( see `getPrivateKeyEntries` below for an example ) and only remove entries if the PrivateKeyEntries are more than 1

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:78]: We definitely need a test for this behavior. One in StoreKeyConfigTests that checks that the created KeyManager has the properties we want it to have

**Review by jkakavas — CHANGES_REQUESTED** [review]
Can we also please change the PR title to something more appropriate and add a PR description that explains why we're adding this ?

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (BigPandaToo) [reply client/rest-high-level/qa/ssl-enabled/build.gradle:32]: Rest Client has a method to read from resources of the class (see `findTrustStore()`), so I put httpCa there for client to find. I could have another copy of httpCa under `qa`, but it would create a potential problem if one copy is replaced in future. Alternatively, I can put transport.p12 under resources and read both files from there - not sure it would look better

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (BigPandaToo) [reply client/rest-high-level/qa/ssl-enabled/build.gradle:26]: Removed

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (BigPandaToo) [reply client/rest-high-level/qa/ssl-enabled/build.gradle:39]: Changed it to `certificate`. The reason I didn't leave it to be full - this transport cert doesn't have SAN (which is required for the full verification mode). I think it is orthogonal to these test purposes.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:83]: Right, good catch, leftover from a previous iteration. Removed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:85]: Added tests

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:78]: Added tests

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [reply client/rest-high-level/qa/ssl-enabled/build.gradle:31]: > Rest Client has a method to read from resources of the class (see findTrustStore()), so I put httpCa there for client to find.

findTruststore is not a RestClient method though right? We add it here in this test, and it basically just returns a `Path`, which we then use in the settings that RestClient is built with. In that sense we can put the file anywhere and as long as our findTruststore returns the Path properly, the RestClient will read this from the settings object. 

>  I can put transport.p12 under resources and read both files from there

That was my suggestion too.

Keeping both files at the same place ensures that we don't miss changing one of them if need arises, etc. Optimally, we'd also have a README.asciidoc in the same dir as the files, to describe how these resources are created if we need to recreate or change them in the future

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [reply client/rest-high-level/qa/ssl-enabled/build.gradle]: > Changed it to certificate.

This is what we will be configuring nodes  with too, so it makes sense to have this to `certificate` instead of `none`. 

> The reason I didn't leave it to be full - this transport cert doesn't have SAN (which is required for the full verification mode).I think it is orthogonal to these test purposes.

We are using this cluster for Enrollment IT and our enrollment process will generate configuration that sets the transport verification mode to `certificate` so it's in our best interest to keep the test cluster config as close to what users will have as possible

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [reply client/rest-high-level/qa/ssl-enabled/build.gradle:34]: maybe just add a TODO here to remind us ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:84]: if you'd filter only key entries before you check for the alias list size, then you don't have to iterate the list in the case where a keystore contains i.e. a PrivateKeyEntry and a TrustedCertificateEntry. As we said, we only want to be doing filtering _iff_ we have more than one PrivateKeyEntry objects. This code will be doing filtering even when we have one PrivateKeyEntry and something else. 

This might be more efficient to do with a stream , while keeping the readability.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (jkakavas) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:72]: I think we can remove this assertion, it doesn't test something that we are interested in ( the certificate chain length of the non-ca certificate we happened to use in the test )

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (jkakavas) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:59]: I'd add load the keystore once here and assert that it has two aliases, and then use their names to do `keyManager.getPrivateKey()` below, so that we are certain that the keymanager filtered out the aliases it should

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (jkakavas) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:91]: maybe quickly add one that contains a keystore with private key entries and trusted certificate entries and check the behavior there too  ?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlMetadataCommandTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlMetadataCommandTests.java:572]: FWIW I'd be fine with us changing the certificates in these tests instead of muting them. Supporting CA certificates was never an intentional thing and there is little value in randomly testing for it. If we want to test that CA certificates can be used for other things than HTTP TLS , let's add specific tests for it to be sure. (In a follow up that we can track in an issue that is )

**Review by jkakavas — CHANGES_REQUESTED** [review]
Thanks for the iteration @BigPandaToo . Let's fix the keystore checking for multiple entries and a couple minor things and it'd be good to be merged

**Comment by jkakavas** [comment]
Suggestion for PR Title : "Filter out CA PrivateKeyEntry when creating a KeyManager" 

Suggestion for the description: ( It'd be nice to have diffs and suggested text as with code :/ but we don't so I add this below , feel free to pick any of the wording you might like ) 

```
In 8.0, with security on by default, we store ca
Private Key Entry used for node and Kibana
enrollment in http.sll keystore. This  keystore is
used for TLS. The current behavior is that the
KeyManager will iterate through the private key
entry and will return the first key/certificate it sees
fits, which lacking any additional logic/filters is any
of the PrivateKeyEntry. If it picks up ca Private
Key instead of the node' one TLS handshake will
fail. The solution is, when creating an in-memory
key store from filter our ca Private Key entry if

    this is http key store;
    there are more than 1 alises in the store (cover
    the case when ca is the node certificate).
```

-----> 

```
In 8.0, with security on by default, we store the HTTP
layer CA PrivateKeyEntry in the  http.ssl keystore (along
with the node certificate) so that it is available in our 
Enrollment API transport actions. 
When loading a keystore, the current behavior is that the
X509ExtendedKeyManager will iterate through the PrivateKeyEntry
objects and will return the first key/certificate that satisfies
the requirements of the client and the server configuration, 
and lacks any additional logic/filters.
We need the KeyManager to deterministically pick the node
certificate/key in all cases as this is the intended entry to be
used for TLS on the HTTP layer. 
This change introduces filtering when creating the in-memory
keystore the KeyManager is loaded with, so that it will not 
include PrivateKeyEntry objects when: 

- there are more than 1 PrivateKeyEntry objects in the keystore 
- The leaf certificate associated with the PrivateKeyEntry is a
CA certificate
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:84]: Good point, changed to checking for the keys first.
Don't like it with stream though, the exception handling looks ugly (isKeyEntry() can throw)

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:72]: Just making sure we haven't delete something we didn't mean to (like it happened before the last chage)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlMetadataCommandTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlMetadataCommandTests.java:572]: Nevertheless, thanks to those tests we discovered potential breaking change we may introduce, I would leave them as they are.

Comment (jkakavas) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlMetadataCommandTests.java:572]: Yeap, but it was just luck that we did discover them because someone at some point created a CA certificate for SAML tests (accidentally). What I am suggesting is to actually test for it since we want to support it and we now know it

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (jkakavas) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:72]: This the certificate chain length of a single certificate chain of a single PrivateKeyEntry. ( `final X509Certificate[] certificates = keyManager.getCertificateChain("http");` ) 

We wouldn't have deleted one of the certificates in that chain before the last change.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:108]: What do you mean? A keystore with multiple not ca private key? Making sure nothing being deleted?

Comment (jkakavas) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:108]: A keystore with a TrustedCertificate entry, a CA PrivateKeyEntry and a non-ca PrivateKeyEntry and make sure that we removed the CA PrivateKeyEntry only

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (jkakavas) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:89]: I still think we don't need this, but won't block the PR on this

File: client/rest-high-level/qa/ssl-enabled/build.gradle
Comment (jkakavas) [reply client/rest-high-level/qa/ssl-enabled/build.gradle:31]: > Optimally, we'd also have a README.asciidoc in the same dir as the files, to describe how these resources are created if we need to recreate or change them in the future

Just repeating this in case you missed it.

**Review by jkakavas — APPROVED** [review]
I have a couple of outstanding comments for your consideration but I don't want to hold the PR for more. See if you can address these if you agree, LGTM otherwise.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlMetadataCommandTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlMetadataCommandTests.java:572]: I will add dedicated tests once we have the fix to limit filtering to tls keystore only

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:108]: I believe that's what httpCa.p12 is in `testCreateKeyManagerFromPKCS12ContainingCA`

Comment (jkakavas) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:108]: no, it is not. it has 2 entries https://github.com/elastic/elasticsearch/pull/73807/files#diff-23d5d8e4b605d268b1ebfad0592f0d6b874c2a3681e9a3213e8b6577388398edR80, i'm describing something that has 3 entries ( A keystore with a TrustedCertificate entry, a CA PrivateKeyEntry and a non-ca PrivateKeyEntry )

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java
Comment (jkakavas) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfigTests.java:75]: we should not be filtering here. We should check that the aliases of all entries are 3 so that we can verify below that we do not remove non PrivateKeyEntry entries

**Pull request merged**
